src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/DocumentPullDiagnosticsHandler.cs

```diff
@@ -0,0 +1,206 @@
+            {
+                return;
+            }
+
+            var remappedResults = await RemapDocumentDiagnosticsAsync(
+                result,
+                razorDocumentUri,
+                documentSnapshot,
+                cancellationToken).ConfigureAwait(false);
+
+            progress.Report(remappedResults);
+        }
+
+        private async Task<DiagnosticReport[]> RemapDocumentDiagnosticsAsync(
+            DiagnosticReport[] unmappedDiagnosticReports,
+            Uri razorDocumentUri,
+            LSPDocumentSnapshot documentSnapshot,
+            CancellationToken cancellationToken)
+        {
+            if (unmappedDiagnosticReports?.Length == 0)
+            {
+                return unmappedDiagnosticReports;
+            }
+
+            var mappedDiagnosticReports = new List<DiagnosticReport>(unmappedDiagnosticReports.Length);
+
+            foreach (var diagnosticReport in unmappedDiagnosticReports)
+            {
+                if (diagnosticReport?.Diagnostics?.Length == 0)
+                {
+                    mappedDiagnosticReports.Add(diagnosticReport);
+                    continue;
+                }
+
+                var unmappedDiagnostics = diagnosticReport.Diagnostics;
+                var mappedDiagnostics = new List<Diagnostic>(unmappedDiagnostics.Length);
+
+                var rangesToMap = unmappedDiagnostics.Select(r => r.Range).ToArray();
+                var mappingResult = await _documentMappingProvider.MapToDocumentRangesAsync(
+                    RazorLanguageKind.CSharp,
+                    razorDocumentUri,
+                    rangesToMap,
+                    LanguageServerMappingBehavior.Inclusive,
+                    cancellationToken).ConfigureAwait(false);
+
+                if (mappingResult == null || mappingResult.HostDocumentVersion != documentSnapshot.Version)
+                {
+                    // Couldn't remap the range or the document changed in the meantime. Discard this highlight.
+                    return Array.Empty<DiagnosticReport>();
+                }
+
+                for (var i = 0; i < unmappedDiagnostics.Length; i++)
+                {
+                    var diagnostic = unmappedDiagnostics[i];
+                    var range = mappingResult.Ranges[i];
+                    if (range.IsUndefined())
+                    {
+                        // Couldn't remap the range correctly. Discard this range.
+                        continue;
+                    }
+
+                    diagnostic.Range = range;
+                    mappedDiagnostics.Add(diagnostic);
+                }
+
+                diagnosticReport.Diagnostics = mappedDiagnostics.ToArray();
+                mappedDiagnosticReports.Add(diagnosticReport);
+            }
+
+            return mappedDiagnosticReports.ToArray();
+        }
+
+        [DataContract]
+        private class SerializableDocumentDiagnosticsParams : DiagnosticParams
```

Comment:
Should this provided at the platform layer?

Reply:
This is due to the `IProgress<T>` serialization bug. Similar to the `SerializableFindAllReferencesParams` that we also have, this class/DataContract is here specifically to ensure we can appropriately serialize our `IProgress<T>` token.